File: output/GPT4o-Mini/xCodeEval_4000/C++/C/1533c622f4378455b3c8b9991c2f60e0.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#define MAXN 100000
#define pii struct PairInt
#define vpii struct VectorPairInt

struct PairInt {
    int X, Y;
};

struct VectorPairInt {
    struct PairInt *data;
    int size;
};

void init_vector_pair_int(struct VectorPairInt *v, int size) {
    v->data = (struct PairInt *)malloc(size * sizeof(struct PairInt));
    v->size = size;
}

void free_vector_pair_int(struct VectorPairInt *v) {
    free(v->data);
}

int raz(int a, int b, int M) {
    int c = a - b;
    return c < 0 ? c + M : c;
}

int sum(int a, int b, int M) {
    int c = a + b;
    return c >= M ? c - M : c;
}

struct PairInt unite_seg(struct PairInt p1, struct PairInt p2, int n) {
    if ((p1.X == -1) || (p2.X == -1))
        return (struct PairInt){-1, -1};
    struct PairInt np1 = {0, raz(p1.Y, p1.X, n)};
    struct PairInt np2 = {raz(p2.X, p1.X, n), raz(p2.Y, p1.X, n)};
    if (np2.X <= np2.Y) {
        if (np2.Y == n - 1)
            return (struct PairInt){-1, -1};
        else
            return (struct PairInt){sum(np1.X, p1.X, n), sum(max(np1.Y, np2.Y), p1.X, n)};
    } else {
        if (np2.X <= np1.Y + 1)
            return (struct PairInt){-1, -1};
        else
            return (struct PairInt){sum(np2.X, p1.X, n), sum(max(np1.Y, np2.Y), p1.X, n)};
    }
}

struct Sparse {
    int n;
    struct VectorPairInt *sp;
};

void construct_sparse(struct Sparse *s, struct VectorPairInt *v) {
    s->n = v->size;
    int r = (int)(log2(s->n)) + 1;
    s->sp = (struct VectorPairInt *)malloc(r * sizeof(struct VectorPairInt));
    for (int i = 0; i < r; i++) {
        init_vector_pair_int(&s->sp[i], s->n);
    }
    for (int i = 0; i < s->n; i++)
        s->sp[0].data[i] = v->data[i];
    for (int i = 1; i < r; i++) {
        for (int j = 0; j < s->n; j++) {
            struct PairInt p1 = s->sp[i - 1].data[j];
            struct PairInt p2 = s->sp[i - 1].data[(j + (1 << (i - 1))) % s->n];
            s->sp[i].data[j] = unite_seg(p1, p2, s->n);
        }
    }
}

struct PairInt segment(struct Sparse *s, int l, int r) {
    if ((l == -1) || (r == -1))
        return (struct PairInt){-1, -1};
    int szz = raz(r, l, s->n);
    if (l == r)
        return s->sp[0].data[l];
    int g = (int)(log2(szz));
    return unite_seg(s->sp[g].data[l], s->sp[g].data[((r + 1 - (1 << g)) % s->n + s->n) % s->n], s->n);
}

int main() {
    int n;
    scanf("%d", &n);
    if (n == 1) {
        printf("0\n");
        return 0;
    }
    int *r = (int *)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        scanf("%d", &r[i]);
    }
    int zz = (int)(log2(n)) + 1;
    struct Sparse *v = (struct Sparse *)malloc(zz * sizeof(struct Sparse));
    struct VectorPairInt prs;
    init_vector_pair_int(&prs, n);

    for (int i = 0; i < n; i++) {
        if (r[i] >= n - 1)
            prs.data[i] = (struct PairInt){-1, -1};
        else
            prs.data[i] = unite_seg((struct PairInt){raz(i, r[i], n), i}, (struct PairInt){i, sum(i, r[i], n)}, n);
    }

    construct_sparse(&v[0], &prs);

    for (int i = 1; i < zz; i++) {
        struct VectorPairInt new_prs;
        init_vector_pair_int(&new_prs, n);
        for (int j = 0; j < n; j++) {
            struct PairInt pairr = segment(&v[i - 1], j, j);
            new_prs.data[j] = segment(&v[i - 1], pairr.X, pairr.Y);
        }
        construct_sparse(&v[i], &new_prs);
        free_vector_pair_int(&new_prs);
    }

    for (int i = 0; i < n; i++) {
        struct PairInt my_pr = {i, i};
        int ans = 0;
        for (int j = zz - 1; j >= 0; --j) {
            struct PairInt new_pr = segment(&v[j], my_pr.X, my_pr.Y);
            if (new_pr.X != -1) {
                ans += (1 << j);
                my_pr = new_pr;
            }
        }
        printf("%d ", ans + 1);
    }
    printf("\n");

    for (int i = 0; i < zz; i++) {
        free_vector_pair_int(&v[i].sp[0]);
        free(v[i].sp);
    }
    free(v);
    free(r);
    free_vector_pair_int(&prs);
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/1533c622f4378455b3c8b9991c2f60e0.py
import sys
from collections import defaultdict
from math import sqrt

def count_tro(r):
    ans = 0
    while r:
        ans += 1
        r >>= 1
    return ans

def unite_seg(p1, p2, n):
    if p1[0] == -1 or p2[0] == -1:
        return (-1, -1)
    np1 = (0, (p1[1] - p1[0]) % n)
    np2 = ((p2[0] - p1[0]) % n, (p2[1] - p1[0]) % n)
    if np2[0] <= np2[1]:
        if np2[1] == n - 1:
            return (-1, -1)
        else:
            return ((np1[0] + p1[0]) % n, max(np1[1], np2[1]) + p1[0]) % n
    else:
        if np2[0] <= np1[1] + 1:
            return (-1, -1)
        else:
            return ((np2[0] + p1[0]) % n, max(np1[1], np2[1]) + p1[0]) % n

class Sparse:
    def __init__(self):
        self.n = 0
        self.sp = []

    def construct(self, v):
        self.n = len(v)
        r = count_tro(self.n - 1)
        self.sp = [[(0, 0)] * self.n for _ in range(r)]
        for i in range(self.n):
            self.sp[0][i] = v[i]
        for i in range(1, r):
            for j in range(self.n):
                p1 = self.sp[i - 1][j]
                p2 = self.sp[i - 1][(j + (1 << (i - 1))) % self.n]
                self.sp[i][j] = unite_seg(p1, p2, self.n)

    def segment(self, l, r):
        if l == -1 or r == -1:
            return (-1, -1)
        szz = r - l
        if l == r:
            return self.sp[0][l]
        g = count_tro(szz) - 1
        return unite_seg(self.sp[g][l], self.sp[g][(r + 1 - (1 << g)) % self.n], self.n)

def main():
    input = sys.stdin.read
    data = input().split()
    n = int(data[0])
    if n == 1:
        print("0")
        return
    r = list(map(int, data[1:n + 1]))
    zz = count_tro(n) + 1
    v = [Sparse() for _ in range(zz)]
    prs = [(0, 0)] * n

    for i in range(n):
        if r[i] >= n - 1:
            prs[i] = (-1, -1)
        else:
            prs[i] = unite_seg((i - r[i]) % n, (i + r[i]) % n, n)

    v[0].construct(prs)

    for i in range(1, zz):
        new_prs = [(0, 0)] * n
        for j in range(n):
            pairr = v[i - 1].segment(j, j)
            new_prs[j] = v[i - 1].segment(pairr[0], pairr[1])
        v[i].construct(new_prs)

    for i in range(n):
        my_pr = (i, i)
        ans = 0
        for j in range(zz - 1, -1, -1):
            new_pr = v[j].segment(my_pr[0], my_pr[1])
            if new_pr[0] != -1:
                ans += (1 << j)
                my_pr = new_pr
        print(ans + 1, end=' ')
    print()

if __name__ == "__main__":
    main()
